### DISCUSSION TITLE
Editing GRASS Commands with Python (r.walk)

### DISCUSSION THREAD

**Discussion by ggh-tech** [OSGeo/grass#3088] [Q&A]
Hello, sorry for this!
 For my dissertation, I have to conduct a Least Cost Route analysis to create a  'most likely' road network. I am hoping to use r.walk in order to include a friction map to bias the network towards larger amounts of sites ( I have created a friction map). However, as I am trying to create a network, I have a single GIS layer of dozens of points (vertexes of roads) to use as my start and stop locations. The problem is that the list is exactly the same and so I am afraid each point will counter itself as the 'nearest stop point to the start point' if you know what I mean when running the analysis in r.walk. As there are so many points it is impractical to create a separate list of points to exclude every individual starting location, and so I was wondering if there was a way to edit the command or find a way around the analysis that each 'starting point' can ignore itself in the 'stop point' list and find the next nearest point? 
I really need this to work and hope for any advice.
Thank you for the time reading!
Nicholas.

**Comment by ggh-tech** [comment]
> Hello, sorry for this! For my dissertation, I have to conduct a Least Cost Route analysis to create a 'most likely' road network. I am hoping to use r.walk in order to include a friction map to bias the network towards larger amounts of sites ( I have created a friction map). However, as I am trying to create a network, I have a single GIS layer of dozens of points (vertexes of roads) to use as my start and stop locations. The problem is that the list is exactly the same and so I am afraid each point will counter itself as the 'nearest stop point to the start point' if you know what I mean when running the analysis in r.walk. As there are so many points it is impractical to create a separate list of points to exclude every individual starting location, and so I was wondering if there was a way to edit the command or find a way around the analysis that each 'starting point' can ignore itself in the 'stop point' list and find the next nearest point? I really need this to work and hope for any advice. Thank you for the time reading! Nicholas.

Sorry, to clarify I am using r.walk in combination with r.drain (hence the need for the start and stop points).

**Reply by petrasovaa** [reply]
Not sure I understand completely, but do you need to use the stop_points? You could run r.walk with each of your points separately as the starting point and then r.drain with all your points, then I suppose you will get multiple paths where one of them will have length 0 (start = stop). Maybe if you show what you have tried, it would help us analyze the problem better. It sounds like little bit of scripting may be needed.

**Reply by ggh-tech** [reply]
Hello So this is what I have done in terms of r.walk with only 4 test points when I likely have  dozens of points.
The analysis stays stuck in the last position so I fear it may be that the DEM is to high a resolution (10m) as well. Thgis is because it never gets past 'Finding cost path'. However, for this analysis, does r.walk and r.drain in combination create paths other than to the nearest point, or is their a way to ignore paths that equate to 0m?

<img width="1920" alt="Screenshot 2023-07-17 154407" src="https://github.com/OSGeo/grass/assets/139667019/ad184f65-bdbd-486c-b859-04f7c51b33dc">
<img width="1920" alt="Screenshot 2023-07-17 155318" src="https://github.com/OSGeo/grass/assets/139667019/5ae8b20a-993e-4c5f-bfc5-d7c6a573e9b9">
<img width="1920" alt="Screenshot 2023-07-17 155325" src="https://github.com/OSGeo/grass/assets/139667019/b7317fc7-a8c1-436f-bc22-2e291949f53b">
<img width="1920" alt="Screenshot 2023-07-17 155333" src="https://github.com/OSGeo/grass/assets/139667019/b829711f-3e33-45b8-8800-fdfa1dfeb7ba">
<img width=
<img width="1920" alt="Screenshot 2023-07-17 155913" src="https://github.com/OSGeo/grass/assets/139667019/6823f0e5-fe24-460d-be5c-1d96082c8639">
"1920" alt="Screenshot 2023-07-17 155359" src="https://github.com/OSGeo/grass/assets/139667019/b200b3e5-e842-42fd-9396-aedd3102cd0d">

**Reply by ggh-tech** [reply]
<img width="1920" alt="Screenshot 2023-07-17 155359" src="https://github.com/OSGeo/grass/assets/139667019/d98d09c0-b8b0-463e-ac1f-32ee485dd8df">

**Reply by ggh-tech** [reply]
is there a way to script it to create several files excluding each point and then run them through r.walk and r.drain?

**Reply by petrasovaa** [reply]
See this script:
https://gis.stackexchange.com/a/173591/22362

It computes a path between each 2 points (plus read the explanation there). If that's not what you need, you need to clarify. If it runs too long, perhaps check computational region (`g.region -p`) and set lower resolution, at least for testing.

BTW, it's better to share the command as text, not several screenshots, the tool's dialog has a copy button for that.

**Comment by ggh-tech** [comment]
Ah alright thankyou, I’m quite new to this, sorry for the trouble

**Reply by petrasovaa** [reply]
No problem

**Comment by ggh-tech** [comment]
Hello I was looking at the script and what i want is a little different,
what I want after messing with r.walk is a way to get a list of points and
then run through each point. for each point analyzed, i want to identify
the closest point (out of all the other points in the list) to the point
under analysis and conduct a least cost path analysis between those two
points.
sorry, I believed r.walk might already have had this feature installed.?
Is what I am talking about possible?

**Reply by petrasovaa** [reply]
You can run r.walk with 1 start point and then all the points (including the start) as stop_points. Export also the direction output. Then you could use either v.what.rast or r.what to query the cumulative cost result at the points, select the closes point and then run r.path. Or, you can run r.path with all the stop_points and then upload the line length to database (v.to.db), and v.extract the 2nd shortest line  (I assume the shortest would be from the start point). And then do the same for all points.
